src/nbl/asset/utils/CIESProfile.cpp

```diff
@@ -0,0 +1,300 @@
+    viewParams.format = viewParams.image->getCreationParameters().format;
+    viewParams.subresourceRange.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+    viewParams.subresourceRange.levelCount = viewParams.image->getCreationParameters().mipLevels;
+    viewParams.subresourceRange.layerCount = 1u;
+
+    return ICPUImageView::create(std::move(viewParams));
+}
+
+//! Explicit instantiations
+template core::smart_refctd_ptr<asset::ICPUImageView> CIESProfile::createIESTexture(const std::execution::sequenced_policy&, const size_t&, const size_t&) const;
+// template core::smart_refctd_ptr<asset::ICPUImageView> CIESProfile::createIESTexture(const std::execution::parallel_policy&, const size_t&, const size_t&) const;
+
+core::smart_refctd_ptr<asset::ICPUImageView> CIESProfile::createIESTexture(const size_t& width, const size_t& height) const
+{
+    return createIESTexture(std::execution::seq, width, height); 
+}
+
+template<class ExecutionPolicy>
+bool CIESProfile::flattenIESTexture(ExecutionPolicy&& policy, const IES_STORAGE_FORMAT& avgEmmision, core::smart_refctd_ptr<asset::ICPUImageView> inoutIES, const float flatten)
+{
+    if constexpr (policy._Parallelize)
+        static_assert(false, "meh it appears that filters with par policy are broken on this branch, use seq policy till fixed");
+
+    const bool inFlattenDomain = flatten >= 0.0 && flatten < 1.0; // [0, 1) range for blend equation, 1 is invalid
+
+    if (!inFlattenDomain)
+        return false;
+
+    auto outIES = inoutIES->getCreationParameters().image;
+
+    const auto& creationParams = outIES->getCreationParameters();
+    const bool validFormat = creationParams.format == CIESProfile::IES_TEXTURE_STORAGE_FORMAT;
+
+    if (!validFormat)
+        return false;
+
+    if (flatten > 0.0) // skip if 0 because its just copying texels
+    {
+        CFillImageFilter::state_type state;
+        state.outImage = outIES.get();
+        state.subresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+        state.subresource.baseArrayLayer = 0u;
+        state.subresource.layerCount = 1u;
+        state.outRange.extent = creationParams.extent;
+
+        const IImageFilter::IState::ColorValue::WriteMemoryInfo wInfo(creationParams.format, outIES->getBuffer()->getPointer());
+        const IImageFilter::IState::ColorValue::ReadMemoryInfo rInfo(creationParams.format, outIES->getBuffer()->getPointer());
+
+        auto fill = [&](uint32_t blockArrayOffset, core::vectorSIMDu32 position) -> void
+        {
+            state.fillValue.readMemory(rInfo, blockArrayOffset);
+
+            const auto decodeV = (IES_STORAGE_FORMAT)(*state.fillValue.asUShort) / CIESProfile::UI16_MAX_D;
+            const auto blendV = decodeV * (1.0 - flatten) + avgEmmision * flatten; //! blend the IES texture with "flatten"
+            const uint16_t encodeV = static_cast<uint16_t>(std::clamp(blendV * CIESProfile::UI16_MAX_D, 0.0, CIESProfile::UI16_MAX_D));
+            
```

Comment:
don't we have `encode` and `decode` functions for this format in `asset` namespace?

Reply:
we do, will use them instead

Reply:
resolved in following commit